pkg/indexer/blockchain/rpcLogStreamer.go

```diff
@@ -0,0 +1,167 @@
+package blockchain
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"go.uber.org/zap"
+)
+
+const (
+	BACKFILL_BLOCKS = 1000
+	// Don't index very new blocks to account for reorgs
+	// Setting to 0 since we are talking about L2s with low reorg risk
+	LAG_FROM_HIGHEST_BLOCK = 0
+	ERROR_SLEEP_TIME       = 100 * time.Millisecond
+)
+
+// The builder that allows you to configure contract events to listen for
+type RpcLogStreamBuilder struct {
+	// All the listeners
+	contractConfigs []contractConfig
+	logger          *zap.Logger
+	rpcUrl          string
+}
+
+func NewRpcLogStreamBuilder(rpcUrl string, logger *zap.Logger) *RpcLogStreamBuilder {
+	return &RpcLogStreamBuilder{rpcUrl: rpcUrl, logger: logger}
+}
+
+func (c *RpcLogStreamBuilder) ListenForContractEvent(fromBlock int, contractAddress common.Address, topics []common.Hash) <-chan types.Log {
+	eventChannel := make(chan types.Log)
+	c.contractConfigs = append(c.contractConfigs, contractConfig{fromBlock, contractAddress, topics, eventChannel})
+	return eventChannel
+}
+
+func (c *RpcLogStreamBuilder) Build() (*RpcLogStreamer, error) {
+	client, err := ethclient.Dial(c.rpcUrl)
+	if err != nil {
+		return nil, err
+	}
+	return NewRpcLogStreamer(client, c.logger, c.contractConfigs), nil
+}
+
+// Struct defining all the information required to filter events from logs
+type contractConfig struct {
+	fromBlock       int
+	contractAddress common.Address
+	topics          []common.Hash
+	channel         chan<- types.Log
+}
+
+/*
+*
+A RpcLogStreamer is a naive implementation of the ChainStreamer interface.
+It queries a remote blockchain node for log events to backfill history, and then streams new events,
+to get a complete history of events on a chain.
+*
+*/
+type RpcLogStreamer struct {
+	client     ChainClient
+	watchers   []contractConfig
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	logger     *zap.Logger
+}
+
+func NewRpcLogStreamer(client ChainClient, logger *zap.Logger, watchers []contractConfig) *RpcLogStreamer {
+	return &RpcLogStreamer{
+		client:   client,
+		watchers: watchers,
+		logger:   logger,
+	}
+}
+
+func (r *RpcLogStreamer) Start(ctx context.Context) error {
+	r.ctx, r.cancelFunc = context.WithCancel(ctx)
+
+	for _, watcher := range r.watchers {
+		go r.watchContract(watcher)
+	}
+	return nil
+}
+
+func (r *RpcLogStreamer) watchContract(watcher contractConfig) {
+	fromBlock := int(watcher.fromBlock)
+	logger := r.logger.With(zap.String("contractAddress", watcher.contractAddress.Hex()))
+	for {
```

Comment:
This is an infinite poll loop without any interval between polls right? Curious if there's a rationale for having no interval?

Reply:
Good call out. I can make it sleep once it gets to the point of having no new logs